tools/mobidrm.1.in
.Dd February 16, 2022
.Dt mobidrm 1 URM
.Os Unix
.Sh NAME
.Nm mobidrm
.Nd Utility for decryption and encryption of MOBI format ebook files.
.Sh SYNOPSIS
.Nm
.Fl d
.Op Fl o Ar dir
.Op Fl p Ar pid
.Op Fl s Ar serial
.Ar file
.Nm
.Fl e
.Op Fl o Ar dir
.Op Fl s Ar serial
.Op Fl f Ar date
.Op Fl t Ar date
.Ar file
.Nm
.Op Fl hv
.Sh DESCRIPTION
The program encrypts or decrypts MOBI files with DRM schemes used on eInk devices.
.Pp
It may encrypt documents targeting particular device by using its serial number or applying generic
DRM scheme targeting all supported readers. Encryption may be valid only during chosen period.
In case of hybrid files mobidrm separately encrypts both enclosed parts and produces two files: mobi for legacy format and azw3 for KF8 format.
.Pp
The program also decrypts documents. One must supply device serial number or document PID if DRM was applied for specific device.
.Pp
Invoked without arguments prints usage summary and exits.
.Pp
A list of flags and their descriptions.
.Pp
\fBDecryption flags\fR:
.Pp
.Bl -tag -width -indent
.It Fl d
Decrypt. This flag must be set in order to decrypt document.
.It Fl p Ar pid
Set optional decryption
.Ar pid
(may be specified multiple times).
.It Fl s Ar serial
Set optional device
.Ar serial
number (may be specified multiple times).
.El
.Pp
\fBEncryption flags\fR:
.Bl -tag -width -indent
.It Fl e
Encrypt. This flag must be set in order to encrypt document.
.It Fl f Ar date
Set optional validity period \fBfrom\fR
.Ar date
formatted as yyyy-mm-dd. Document will not be valid before this date.
.It Fl s Ar serial
Set optional device
.Ar serial
number. If not set document will be encrypted with generic reader key. This flag may be specified
multiple times to target multiple devices.
.It Fl t Ar date
Set optional validity period \fBto\fR
.Ar date
formatted as yyyy-mm-dd. Document will expire after this date.
.El
.Pp
\fBCommon flags\fR:
.Bl -tag -width -indent
.It Fl o Ar dir
Save output to
.Ar dir
folder.
.It Fl h
Show usage summary and exit.
.It Fl v
Show version and exit.
.El
.Pp
.Sh EXAMPLES
The following command decrypts document using given serial number. Decrypted output will be saved to /tmp folder.
.Pp
.Bl -item -offset indent
.Nm
.Fl d
.Fl s
B001XXXXXXXXXXXX
.Fl o
/tmp example.mobi
.El
.Pp
The following command decrypts document trying multiple PIDs.
.Pp
.Bl -item -offset indent
.Nm
.Fl d
.Fl p
XXXXXXXXX1
.Fl p
XXXXXXXXX2 example.mobi
.El
.Pp
The following command encrypts document targeting two devices using their serial numbers. Document will be readable only on those devices.
.Pp
.Bl -item -offset indent
.Nm
.Fl e
.Fl s
B001XXXXXXXXXX01
.Fl s
B001XXXXXXXXXX02 example.mobi
.El
.Pp
The following command encrypts document for device wth given serial number. Document will only be valid within given period (inclusive).
.Pp
.Bl -item -offset indent
.Nm
.Fl e
.Fl s
B001XXXXXXXXXX01
.Fl f
2021-01-01
.Fl t
2021-01-31 example.mobi
.El
.Sh RETURN VALUES
The
.Nm
utility returns 0 on success, 1 on error.
.Sh COPYRIGHT
Copyright (C) 2021-2022 Bartek Fabiszewski.
.Pp
Released under LGPL version 3 or any later (same as libmobi).
.Sh WEB SITE
Visit http://www.fabiszewski.net for details.
.Sh DIAGNOSTICS
For diagnostics
.Nm libmobi
must be configured with
.Op Fl Fl enable-debug
option.
.Sh SEE ALSO
.Xr mobimeta 1
.Xr mobitool 1


ChangeLog
2022-02-25: Remove obsolete changelog
2022-02-25: Fix md5sum output on Windows
2022-02-25: Fix inconsistent separators in path on Windows builds
2022-02-25: GHA: fix log paths
2022-02-25: GHA: fix workflow syntax
2022-02-25: GHA: upload test logs on failure
2022-02-24: Fix printf format specifier
2022-02-24: Fix sample path in Makefile
2022-02-24: Missing autotools in mingw workflow
2022-02-24: Windows doesn't accept asterisk in file names
2022-02-24: Update workflow, add badge
2022-02-24: Add mingw workflow
2022-02-24: Fix tests in out-of-tree build
2022-02-24: Update man pages
2022-02-24: Replace non-portable strptime
2022-02-24: Make sure both validity period dates are set
2022-02-21: Fix strptime not found on linux build
2022-02-21: Add build github action
2022-02-21: Update README
2022-02-21: Minor code cleanups
2022-02-21: Unify boolean and static usage in tools
2022-02-21: mobimeta: fix null pointer dereference when parsing malformed option
2022-02-18: Add  hybrid spit option to mobitool
2022-02-18: Update documentation
2022-02-18: Test both encrypted hybrid parts
2022-02-18: Fix: fast decryption routine fails for non-huffman compression
2022-02-18: Fix mobitool serial decryption
2022-02-18: Add DRM tests
2022-02-17: Fix build with encryption disabled
2022-02-17: Update tests samples
2022-02-16: Add -h option to tools, update man pages
2022-02-16: Update Xcode settings
2022-02-16: Restructure, cleanup encryption related code, add mobidrm tool
2021-11-19: Improve getopt loop, fix config.h to be accessible from all tools
2021-11-10: Update xcode project
2021-11-10: Add functions to split hybrid files
2021-11-10: Avoid modifying existing records, as caller may keep reference to them
2021-11-05: Fix: tests fail if pid contains asterisk
2021-11-05: Fix: decryption may fail for some records with standard compression
2021-11-05: Replace test samples with self-generated smaller ones
2021-11-05: Skip test in case of missing checksums
2021-10-20: Version 0.9
2021-10-24: Fix out-of-tree build
2021-10-22: Fix mingw build, code formatting
2021-10-14: Fix gcc format truncation warning
2021-10-14: Include autogen.sh in distribution bundle
2021-10-14: Create codeql-analysis.yml
2021-10-14: Fix autoconf 2.70 warnings, clean up
2021-10-14: Build fails with autoconf 2.70
2021-10-11: Version 0.8
2021-10-11: Update Xcode project
2021-10-11: Fix warnings about changed signedness
2021-09-18: Fix potential out-of-buffer read while parsing corrupt file, closes #38
2021-09-18: Fix potential out-of-buffer read while parsing corrupt file, closes #35, #36
2021-09-09: Version 0.7
2021-09-09: fix oob write bug inside libmobi
2021-06-07: Add reference to brew formula
2020-09-02: Fix null pointer dereference in case of broken fragment
2020-08-01: Update changelog
2020-08-01: Version 0.6
2020-07-31: Fix typo
2020-07-31: Add Readme to dist package
2020-07-31: Remove anchor on truncated link
2020-07-31: Fix missing option in man page
2020-07-30: Include test samples in dist package
2020-07-25: Fix gcc 7+ warnings about implicit fall through and format truncation
2020-07-24: Unique names for internal functions to avoid confilicts with static linking
2020-06-24: Close file in error branch
2020-06-24: Fix static compilation with miniz on gcc
2020-06-24: Minor documentation fixes
2020-06-23: Version 0.5
2020-06-23: mobitool: add dump cover option
2020-06-23: Minor documentation improvement
2020-06-23: Fix potential buffer over-read
2019-03-18: Fix: try also "name" attribute when searching for link anchor tags, closes #24
2019-02-22: Add mobi_is_replica function
2019-02-22: Fix potential read beyond buffer
2019-02-22: Travis migration
2018-08-07: Fix: missing items in recreated ncx file
2018-06-20: Fix: printf format warning on some gcc versions
2018-06-20: Fix: make dist broken by nonexistent header files
2018-06-20: VERSION 0.4
2018-06-20: Fix: buffer overflow (CVE-2018-11726)
2018-06-20: Fix: buffer overflow (CVE-2018-11724)
2018-06-20: Fix: read beyond buffer (CVE-2018-11725)
2018-06-20: Fix: buffer overflow (mobitool), closes #18
2018-06-20: Fix: read beyond buffer with corrupted KF8 Boundary record, closes #19
2018-06-20: Fix: read beyond buffer, closes #16, #17
2018-06-20: Updated xcode project files
2018-04-03: Fix: ncx part was not scanned for links, fixes #12
2018-04-02: Fix regression, potential use after free
2018-04-02: Skip broken resources, fixes #10
2018-03-05: Allow processing zero length text records, fixes #9
2017-12-25: Skip broken first resource offset instead of dying
2017-12-18: Skip broken links reconstruction instead of dying
2017-11-27: Disable travis OS X builds, as they usually time out
2017-11-16: Fix: increase max number of dictionary entries per record
2017-11-14: Fix for some encrypted documents with palmdoc encoding
2017-11-06: Fix: potential null pointer dereference
2017-10-16: Manpage cleanup
2017-09-27: Update README
2017-09-26: Increase maximum length of attribute name and value, closes #5
2017-02-26: Remove obsolete files from VS build (closes #3) [ci skip]
2016-11-05: Mobitool: use epub extension if extracted source resource is epub
2016-06-10: Update docs
2016-06-10: Update test files
2016-06-10: Fix: out of bounds read in corrupt font resource
2016-06-10: Prevent memory leak in case of corrupt font resources
2016-06-10: Calculate deobfuscation buffer limit from key length
2016-06-10: Fix: USE_LIBXML2 macro was not included from config.h
2016-06-10: Fix: USE_LIBXML2 macro was not included from config.h
2016-06-09: Fix: memory leak in tools
2016-06-09: Fix: potential out of bounds read
2016-06-09: Fix: memory leak in internal xmlwriter
2016-06-01: Update README
2016-05-19: Feature: verify decryption key type
2016-05-19: Cleanup converting little endian buffer to 32-bit integer
2016-05-19: Feature: check drm expiration dates
2016-05-18: Fix: memory leaks in encryption
2016-05-18: Fix concurrent autotools builds

tools/mobidrm.c
/** @file mobidrm.c
 *
 * @brief mobidrm
 *
 * Copyright (c) 2021 Bartek Fabiszewski
 * http://www.fabiszewski.net
 *
 * Licensed under LGPL, either version 3, or any later.
 * See <http://www.gnu.org/licenses/>
 */

#include <string.h>
#include <stdbool.h>
#include <stdlib.h>
#include <ctype.h>
#include <errno.h>
#include <mobi.h>

#include "common.h"

#define VOUCHERS_COUNT_MAX 20

/* command line options */
bool decrypt_opt = false;
bool encrypt_opt = false;
bool expiry_opt = false;

/* options values */
char *pid[VOUCHERS_COUNT_MAX];
char *serial[VOUCHERS_COUNT_MAX];
size_t serial_count = 0;
size_t pid_count = 0;
time_t valid_from = -1;
time_t valid_to = -1;


/**
 @brief Print usage info
 @param[in] progname Executed program name
 */
static void print_usage(const char *progname) {
    printf("usage: %s [-d | -e] [-hv] [-p pid] [-f date] [-t date] [-s serial] [-o dir] filename\n", progname);
    printf("       without arguments prints document metadata and exits\n\n");

    printf("       Decrypt options:\n");
    printf("       -d        decrypt (required)\n");
    printf("       -p pid    set decryption pid (may be specified multiple times)\n");
    printf("       -s serial set device serial (may be specified multiple times)\n\n");
    
    printf("       Encrypt options:\n");
    printf("       -e        encrypt (required)\n");
    printf("       -s serial set device serial (may be specified multiple times)\n");
    printf("       -f date   set validity period from date (yyyy-mm-dd) when encrypting (inclusive)\n");
    printf("       -t date   set validity period to date (yyyy-mm-dd) when encrypting (inclusive)\n");
    
    printf("       Common options:\n");
    printf("       -o dir    save output to dir folder\n");
    printf("       -h        show this usage summary and exit\n");
    printf("       -v        show version and exit\n");
}

/**
 @brief Applies DRM
 
 @param[in,out] m MOBIData structure
 @param[in] use_kf8 In case of hybrid file process KF8 part if true, the other part otherwise
 @return SUCCESS or ERROR
 */
static int do_encrypt(MOBIData *m, bool use_kf8) {
    
    MOBI_RET mobi_ret;
    
    if (mobi_is_hybrid(m)) {
        mobi_ret = mobi_remove_hybrid_part(m, !use_kf8);
        if (mobi_ret != MOBI_SUCCESS) {
            printf("Error removing hybrid part (%s)\n", libmobi_msg(mobi_ret));
            return ERROR;
        }
        printf("\nProcessing file version %zu from hybrid file\n", mobi_get_fileversion(m));
    }
    
    MOBIExthTag *tags = NULL;
    MOBIExthTag amazonkeys[] = { EXTH_WATERMARK, EXTH_TTSDISABLE, EXTH_CLIPPINGLIMIT, EXTH_READFORFREE, EXTH_RENTAL, EXTH_UNK407 };
    MOBIExthTag tamperkeys[ARRAYSIZE(amazonkeys)] = { 0 };
    size_t tags_count = 0;
    
    if (mobi_get_fileversion(m) >= 4 && m->eh) {
        // EXTH tamperproof keys are required with encryption scheme 2 in mobi version 8
        // and are supported since version 4.
        // Amazon software by default includes tags: 208, 404, 401, 405, 406, 407 in DRM scheme.
        // Any change in tags that are included invalidates DRM.
        // You can secure any tag you want, we just add dummy watermark (208) if it is not already present in file
        // to satisfy requirements.
        if (mobi_get_exthrecord_by_tag(m, EXTH_WATERMARK) == NULL) {
            char watermark[] = "aHR0cHM6Ly9naXRodWIuY29tL2JmYWJpc3pld3NraS9saWJtb2Jp";
            mobi_add_exthrecord(m, EXTH_WATERMARK, (uint32_t) strlen(watermark), watermark);
        }
        tamperkeys[0] = amazonkeys[0];
        tags_count++;
        // If any of the Amazon default tags is present in file, secure it too.
        // You can also add any of them with mobimeta before running mobidrm.
        for (size_t i = 1; i < ARRAYSIZE(amazonkeys); i++) {
            // if exth record exists in mobi file
            if (mobi_get_exthrecord_by_tag(m, amazonkeys[i]) != NULL) {
                // add it tamperproof keys
                tamperkeys[tags_count++] = amazonkeys[i];
            }
        }
        
        tags = tamperkeys;
    }

    for (size_t i = 0; i < serial_count; i++) {
        mobi_ret = mobi_drm_addvoucher(m, serial[i], valid_from, valid_to, tags, tags_count);
        if (mobi_ret != MOBI_SUCCESS) {
            printf("Error adding encryption voucher (%s)\n", libmobi_msg(mobi_ret));
            return ERROR;
        }
    }
    if (serial_count == 0) {
        mobi_ret = mobi_drm_addvoucher(m, NULL, valid_from, valid_to, tags, tags_count);
        if (mobi_ret != MOBI_SUCCESS) {
            printf("Error adding encryption voucher (%s)\n", libmobi_msg(mobi_ret));
            return ERROR;
        }
    }
    
    mobi_ret = mobi_drm_encrypt(m);
    if (mobi_ret != MOBI_SUCCESS) {
        printf("Error encrypting document (%s)\n", libmobi_msg(mobi_ret));
        return ERROR;
    }
    
    printf("Encrypting with encryption type %u\n", m->rh->encryption_type);
    if (m->rh->encryption_type == MOBI_ENCRYPTION_V1 && serial_count) {
        printf("Warning! Encryption with device serial number is not supported with this encryption scheme. Skipping serial...\n");
    }
    if (valid_from != -1) {
        printf("Drm validity period from %s", ctime(&valid_from));
    }
    if (valid_to != -1) {
        printf("Drm validity period to %s", ctime(&valid_to));
    }
    return SUCCESS;
}

/**
 @brief Removes DRM
 
 @param[in,out] m MOBIData structure
 @return SUCCESS or ERROR
 */
static int do_decrypt(MOBIData *m) {
    MOBIExthHeader *exth = mobi_get_exthrecord_by_tag(m, EXTH_RENTAL);
    if (exth) {
        uint32_t is_rental = mobi_decode_exthvalue(exth->data, exth->size);
        if (is_rental) {
            printf("Can't remove DRM from rented documents\n");
            return ERROR;
        }
    }
    uint16_t encryption_type = m->rh->encryption_type;
    printf("Removing encryption type %u\n", encryption_type);
    bool has_key = false;
    if (pid_count) {
        for (size_t i = 0; i < pid_count; i++) {
            if (set_decryption_pid(m, pid[i]) == SUCCESS) {
                has_key = true;
                break;
            }
        }
    }
    if (!has_key && serial_count) {
        for (size_t i = 0; i < serial_count; i++) {
            if (set_decryption_serial(m, serial[i]) == SUCCESS) {
                break;
            }
        }
    }
    MOBI_RET mobi_ret = mobi_drm_decrypt(m);
    if (mobi_ret != MOBI_SUCCESS) {
        printf("Error encrypting document (%s)\n", libmobi_msg(mobi_ret));
        return ERROR;
    }
    
    if (encryption_type == MOBI_ENCRYPTION_V2) {
        // remove EXTH records that impose restrictions or are DRM related
        MOBIExthTag drmkeys[] = { EXTH_TAMPERKEYS, EXTH_WATERMARK, EXTH_TTSDISABLE, EXTH_CLIPPINGLIMIT, EXTH_READFORFREE, EXTH_RENTAL, EXTH_UNK407 };
        for (size_t i = 0; i < ARRAYSIZE(drmkeys); i++) {
            mobi_ret = mobi_delete_exthrecord_by_tag(m, drmkeys[i]);
            if (mobi_ret != MOBI_SUCCESS) {
                printf("Error removing EXTH record %u\n", drmkeys[i]);
                return ERROR;
            }
        }
    }
    return SUCCESS;
}

/**
 @brief Main routine that calls optional subroutines
 
 @param[in] fullpath Full file path
 @return SUCCESS or ERROR
 */
static int loadfilename(const char *fullpath) {

    static int run_count = 0;
    run_count++;
    
    bool use_kf8 = run_count == 1 ? false : true;
    
    /* Initialize main MOBIData structure */
    MOBIData *m = mobi_init();
    if (m == NULL) {
        printf("Memory allocation failed\n");
        return ERROR;
    }
    
    errno = 0;
    FILE *file = fopen(fullpath, "rb");
    if (file == NULL) {
        int errsv = errno;
        printf("Error opening file: %s (%s)\n", fullpath, strerror(errsv));
        mobi_free(m);
        return ERROR;
    }
    
    /* MOBIData structure will be filled with loaded document data and metadata */
    MOBI_RET mobi_ret = mobi_load_file(m, file);
    fclose(file);
    
    if (mobi_ret != MOBI_SUCCESS) {
        printf("Error while loading document (%s)\n", libmobi_msg(mobi_ret));
        mobi_free(m);
        return ERROR;
    }
    
    if (run_count == 1) {
        print_summary(m);
    }

    // save is hybrid result here, as file may be modified later
    bool is_hybrid = mobi_is_hybrid(m);
    
    int ret = SUCCESS;
    const char *suffix;
    if (encrypt_opt) {
        // encrypt
        if (mobi_is_encrypted(m)) {
            mobi_free(m);
            printf("Document is already encrypted\n");
            return ERROR;
        }
        ret = do_encrypt(m, use_kf8);
        if (ret != SUCCESS) {
            mobi_free(m);
            return ERROR;
        }
        suffix = "encrypted";
    } else {
        // decrypt
        if (!mobi_is_encrypted(m)) {
            mobi_free(m);
            printf("Document is not encrypted\n");
            return ERROR;
        }
        ret = do_decrypt(m);
        if (ret != SUCCESS) {
            mobi_free(m);
            return ERROR;
        }
        suffix = "decrypted";
    }
    
    ret = save_mobi(m, fullpath, suffix);
    if (ret != SUCCESS) {
        mobi_free(m);
        return ERROR;
    }
    
    /* Free MOBIData structure */
    mobi_free(m);
    
    /* In case of encrypting hybrid file proceed with KF8 part */
    if (encrypt_opt && is_hybrid && use_kf8 == false) {
        loadfilename(fullpath);
    }
    return SUCCESS;
}

/**
 @brief Parse ISO8601 date into tm structure
 
 @param[in,out] tm Structure to be filled
 @param[in] date_str Input date string
 @return SUCCESS or ERROR
 */
static int parse_date(struct tm *tm, const char *date_str) {
    memset(tm, '\0', sizeof(*tm));
    int year;
    int month;
    int day;
    if (sscanf(date_str, "%5d-%2d-%2d", &year, &month, &day) != 3) {
        return ERROR;
    }
    if (year < 1000 || month < 1 || month > 12 || day < 1 || day > 31) {
        return ERROR;
    }
    
    tm->tm_year = year - 1900;
    tm->tm_mon = month - 1;
    tm->tm_mday = day;
    tm->tm_isdst = -1;
    
    return SUCCESS;
}

/**
 @brief Main
 
 @param[in] argc Arguments count
 @param[in] argv Arguments array
 @return SUCCESS (0) or ERROR (1)
 */
int main(int argc, char *argv[]) {
    if (argc < 2) {
        print_usage(argv[0]);
        return ERROR;
    }
    opterr = 0;
    int c;
    while ((c = getopt(argc, argv, "def:ho:p:s:t:vx:")) != -1) {
        switch(c) {
            case 'd':
                if (encrypt_opt) {
                    printf("Options -d and -e can't be used together.\n");
                    return ERROR;
                }
                decrypt_opt = true;
                break;
            case 'e':
                if (decrypt_opt) {
                    printf("Options -d and -e can't be used together.\n");
                    return ERROR;
                }
                encrypt_opt = true;
                break;
            case 'f':
                if (strlen(optarg) == 2 && optarg[0] == '-') {
                    printf("Option -%c requires an argument.\n", c);
                    return ERROR;
                }
                struct tm from;
                if (parse_date(&from, optarg) != SUCCESS) {
                    printf("Wrong valid from date format, use ISO 8601 yyyy-mm-dd\n");
                    return ERROR;
                }
                valid_from = mktime(&from);
                if (valid_from == -1) {
                    printf("Unparsable valid from date format, use ISO 8601 yyyy-mm-dd\n");
                    return ERROR;  
                }
                expiry_opt = true;
                break;
            case 'o':
                if (strlen(optarg) == 2 && optarg[0] == '-') {
                    printf("Option -%c requires an argument.\n", c);
                    return ERROR;
                }
                size_t outdir_length = strlen(optarg);
                if (outdir_length >= FILENAME_MAX - 1) {
                    printf("Output directory name too long\n");
                    return ERROR;
                }
                strncpy(outdir, optarg, FILENAME_MAX - 1);
                normalize_path(outdir);
                if (!dir_exists(outdir)) {
                    printf("Output directory is not valid\n");
                    return ERROR;
                }
                if (optarg[outdir_length - 1] != separator) {
                    // append separator
                    if (outdir_length >= FILENAME_MAX - 2) {
                        printf("Output directory name too long\n");
                        return ERROR;
                    }
                    outdir[outdir_length++] = separator;
                    outdir[outdir_length] = '\0';
                }
                outdir_opt = true;
                break;
            case 'p':
                if (strlen(optarg) == 2 && optarg[0] == '-') {
                    printf("Option -%c requires an argument.\n", c);
                    return ERROR;
                }
                if (pid_count < VOUCHERS_COUNT_MAX) {
                    pid[pid_count++] = optarg;
                } else {
                    printf("Maximum PIDs count reached, skipping PID...\n");
                }
                break;
            case 's':
                if (strlen(optarg) == 2 && optarg[0] == '-') {
                    printf("Option -%c requires an argument.\n", c);
                    return ERROR;
                }
                if (serial_count < VOUCHERS_COUNT_MAX) {
                    serial[serial_count++] = optarg;
                } else {
                    printf("Maximum serial numbers count reached, skipping serial...\n");
                }
                break;
            case 't':
                if (strlen(optarg) == 2 && optarg[0] == '-') {
                    printf("Option -%c requires an argument.\n", c);
                    return ERROR;
                }
                struct tm to;
                if (parse_date(&to, optarg) != SUCCESS) {
                    printf("Wrong valid to date format, use ISO 8601 yyyy-mm-dd\n");
                    return ERROR;
                }
                to.tm_hour = 23;
                to.tm_min = 59;
                to.tm_sec = 59;
                valid_to = mktime(&to);
                if (valid_to == -1) {
                    printf("Unparsable valid to date format, use ISO 8601 yyyy-mm-dd\n");
                    return ERROR;  
                }
                expiry_opt = true;
                break;
            case 'v':
                printf("mobidrm build: " __DATE__ " " __TIME__ " (" COMPILER ")\n");
                printf("libmobi: %s\n", mobi_version());
                return SUCCESS;
            case '?':
                if (isprint(optopt)) {
                    fprintf(stderr, "Unknown option `-%c'\n", optopt);
                }
                else {
                    fprintf(stderr, "Unknown option character `\\x%x'\n", optopt);
                }
                print_usage(argv[0]);
                return ERROR;
            case 'h':
            default:
                print_usage(argv[0]);
                return SUCCESS;
        }
    }
    
    if (argc <= optind) {
        printf("Missing filename\n");
        print_usage(argv[0]);
        return ERROR;
    }
    
    if (!decrypt_opt && !encrypt_opt) {
        printf("One of -d (decrypt) and -e (encrypt) options is required\n");
        print_usage(argv[0]);
        return ERROR;
    }

    if (!encrypt_opt && expiry_opt) {
        printf("Expiration flags can only be used with -e (encrypt)\n");
        print_usage(argv[0]);
        return ERROR;
    }
    
    if (!decrypt_opt && pid_count) {
        printf("PID can only be used with -d (decrypt)\n");
        print_usage(argv[0]);
        return ERROR;
    }
    
    int ret = 0;
    char filename[FILENAME_MAX];
    strncpy(filename, argv[optind], FILENAME_MAX - 1);
    filename[FILENAME_MAX - 1] = '\0';
    normalize_path(filename);

    ret = loadfilename(filename);
    
    return ret;
}
